pkg/handler/list_owner_nft.go

```diff
@@ -88,6 +74,35 @@ func (h *ListOwnerNFTAPIHandler) ServeHTTP(resp http.ResponseWriter, req *http.R
 		contractIDToCollectionMap[contractID] = collection
 	}
 
+	filteredContracts := make([]authgearweb3.ContractID, 0)
+	for _, url := range urlValues["contract_id"] {
+		e, err := authgearweb3.ParseContractID(url)
+		if err != nil {
+			h.Logger.WithError(err).Error("failed to parse contract ID")
+			h.JSON.WriteResponse(resp, &authgearapi.Response{Error: apierrors.NewBadRequest("invalid contract ID")})
+			return
+		}
+
+		// Check whether the contract is being watched or not
+		if _, ok := contractIDToCollectionMap[*e]; !ok {
+			h.Logger.Error("one or more contract ID(s) are not being watched")
+			h.JSON.WriteResponse(resp, &authgearapi.Response{Error: apierrors.NewBadRequest(fmt.Sprintf("contract %s is not being watched", url))})
+			return
+		}
+
+		// Filter out contracts that are not in owner's network
+		if e.Blockchain == ownerID.Blockchain && e.Network == ownerID.Network {
+			filteredContracts = append(filteredContracts, *e)
+		}
+	}
+
+	// Ensure there are at least one valid contract ID
+	if len(filteredContracts) == 0 {
+		h.Logger.Error("invalid contract ID")
+		h.JSON.WriteResponse(resp, &authgearapi.Response{Error: apierrors.NewBadRequest("missing contract ID")})
+		return
+	}
```

Comment:
The point here is to compute an intersection between the incoming contract IDs and the collections being watched. This can be done with a SQL query, instead of fetching all collections from the database to compute in application code.

```SQL
SELECT *
FROM eth_nft_collection
WHERE 
  (blockchain = $blockchain1 AND network = $network1 AND contract_address = $contract_address1)
  OR (blockchain = $blockchain2 AND network = $network2 AND contract_address = $contract_address2)
  OR ...
```